migration-engine/connectors/sql-migration-connector/src/flavour/mysql.rs

```diff
@@ -119,13 +123,95 @@ impl SqlFlavour for MysqlFlavour {
         Ok(connection.raw_cmd(&query).await?)
     }
 
+    async fn run_query_script(&self, sql: &str, connection: &Connection) -> ConnectorResult<()> {
+        let convert_error = |error: my::Error| match error {
+            my::Error::Server(se) if self.is_vitess() && se.code == 1317 => {
+                ConnectorError::user_facing(ForeignKeyCreationNotAllowed)
+            }
+            _ => quaint_error_to_connector_error(quaint::error::Error::from(error), &connection.connection_info()),
+        };
+
+        let (client, _url) = connection.unwrap_mysql();
+        let mut conn = client.conn().lock().await;
+
+        let mut result = sql.run(&mut *conn).await.map_err(convert_error)?;
+
+        loop {
+            match result.map(drop).await {
+                Ok(_) => {
+                    if result.is_empty() {
+                        result.map(drop).await.map_err(convert_error)?;
+                        return Ok(());
+                    }
+                }
+                Err(e) => {
+                    return Err(convert_error(e));
+                }
+            }
+        }
+    }
+
     async fn apply_migration_script(
         &self,
         migration_name: &str,
         script: &str,
-        conn: &Connection,
+        connection: &Connection,
     ) -> ConnectorResult<()> {
-        super::generic_apply_migration_script(migration_name, script, conn).await
+        let convert_error = |migration_idx: usize, error: my::Error| {
+            let position = format!(
+                "Please check the query number {} from the migration file.",
+                migration_idx
+            );
+
+            let (code, error) = match error {
+                my::Error::Server(se) if self.is_vitess() && se.code == 1317 => {
+                    let error =
+                        "Foreign keys can not be created on this database. Learn more how to handle this: https://pris.ly/d/migrate-no-foreign-keys";
```

Comment:
Could there be a way to share that error handling code between `run_query_script()` and `apply_migration_script()`?

Reply:
I can try. It almost can, and I always find something that breaks my shared logic (or makes it annoying to read)...